dev-support/git-hooks/pre-commit/100-verify-submodules-pushed.sh

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Redirect output to stderr.
+exec 1>&2
+
+
+#set -o xtrace
+set -o errexit
+set -o pipefail
+set -o nounset
+
+bin="$(cd "$(dirname "$0")" > /dev/null; pwd)"
+
+# Status Table
+# A         Added
+# C         Copied
+# D         Deleted
+# M         Modified
+# R         Renamed
+# T         Type Changed (i.e. regular file, symlink, submodule, …<200b>)
+# U         Unmerged
+# X         Unknown
+# B         Broken
+_main() {
+  # In case the usage happens at a different layer, make sure to cd to the toplevel
+  local root_dir
+  root_dir="$(git rev-parse --show-toplevel)"
+  cd "$root_dir"
+
+  if [[ ! -e .gitmodules ]]; then
+    # nothing to see here, look away!
+    return 0
+  fi
+  local enabled=$(git config --bool cassandra.pre-commit.verify-submodules.enabled || echo true)
+  [ "$enabled" == "false" ] && return 0
+  local submodules=( $(git config --file .gitmodules --get-regexp path | awk '{ print $2 }') )
+
+  local is_submodule=false
+  local git_sub_dir
+  while read status file; do
+    is_submodule=false
+    for to_check in "${submodules[*]}"; do
+      if [[ "$to_check" == "$file" ]]; then
+        is_submodule=true
+        break
+      fi
+    done
+    if $is_submodule; then
+      local enabled=$(git config --bool cassandra.pre-commit.verify-submodule-${file}.enabled || echo true)
+      [ "$enabled" == "false" ] && continue
+      echo -e "[pre-commit]\tSubmodule detected: ${file} with status ${status}; attempting a push"
+      git_sub_dir="${file}/.git"
+      branch="$(git --git-dir "${git_sub_dir}" rev-parse --abbrev-ref HEAD)"
+      if [[ "$branch" == "HEAD" ]]; then
+        # can't find the remote... so push all
+        git --git-dir "${git_sub_dir}" push --atomic
+      else
+        default_remote="$(git --git-dir "${git_sub_dir}" config --local --get branch."${branch}".remote || true)"
+        remote="${default_remote:-origin}"
+
+        git --git-dir "${git_sub_dir}" push --atomic "$remote" "$branch"
```

Comment:
not sure about this auto pushing (here and in `pre-push/`)… 
but i'm not entirely clear on the intent of both scripts – can we add comments after the license explaining when and what and why the scripts are doing what they are doing…

Reply:
ill add a comment, but ill also explain here.

If you are doing development in both git projects, you will create a sequence of SHAs.  Now, if you rewrite the accord history, you will have a C* commit pointing to an accord SHA that does not exist.   To make sure all commits can build, we push the accord changes up, so if anyone checks out C* it can build.  

This is why you had C* commits that failed to find a SHA, its because the sha was never pushed up to GitHub and history was rewritten removing it.

Reply:
pushed a comment explaining why this is done, and improved the logging to show users how to disable.

Reply:
I guess another option is to detect that local accord is further ahead than remote and block commit until resolved?  I could always make it a config so you choose which solution you like...

in development I like it pushing for me...